plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/refactoring/N4JSRenameElementHandler.java

```diff
@@ -0,0 +1,135 @@
+import org.eclipse.xtext.resource.XtextResource;
+import org.eclipse.xtext.ui.editor.XtextEditor;
+import org.eclipse.xtext.ui.editor.utils.EditorUtils;
+import org.eclipse.xtext.ui.refactoring.ui.DefaultRenameElementHandler;
+import org.eclipse.xtext.ui.refactoring.ui.IRenameElementContext;
+import org.eclipse.xtext.util.concurrent.IUnitOfWork;
+
+import com.google.inject.Inject;
+
+/**
+ * Customer rename handler
+ */
+@SuppressWarnings("restriction")
+public class N4JSRenameElementHandler extends DefaultRenameElementHandler {
+	@Inject
+	private EObjectAtOffsetHelper eObjectAtOffsetHelper;
+
+	@Override
+	public Object execute(ExecutionEvent event) throws ExecutionException {
+		try {
+			final XtextEditor editor = EditorUtils.getActiveXtextEditor(event);
+			if (editor != null) {
+				syncUtil.totalSync(preferences.isSaveAllBeforeRefactoring(),
+						renameRefactoringController.getActiveLinkedMode() == null);
+				final ITextSelection selection = (ITextSelection) editor.getSelectionProvider().getSelection();
+				IRenameElementContext renameElementContext = editor.getDocument().priorityReadOnly(
+						new IUnitOfWork<IRenameElementContext, XtextResource>() {
+							@Override
+							public IRenameElementContext exec(XtextResource resource) throws Exception {
+								EObject selectedElement = eObjectAtOffsetHelper.resolveElementAt(resource,
+										selection.getOffset());
+
+								// GH-1002: Ask Oliver
+								// Special handling for FormalParameter
+								// IdentifierRef refers to FormalParameter
+								EObject selectedTypeElement = (selectedElement instanceof FormalParameter) ? null
+										: N4JSLanguageUtils
+												.getDefinedTypeModelElement(selectedElement);
+								selectedElement = selectedTypeElement == null ? selectedElement
+										: selectedTypeElement;
+
+								// GH-1002: Do NOT allow renaming alias if the selected element is imported in the
+								// current resource via alias.
+								// This does not work in case the element is alias imported in another resource (this is
+								// expensive to check).
+								N4JSResource res = (N4JSResource) resource;
+								Script script = (Script) res.getContents().get(0);
+								List<ImportDeclaration> importDecls = script.getScriptElements().stream()
+										.filter(elem -> elem instanceof ImportDeclaration)
+										.map(elem -> (ImportDeclaration) elem)
+										.collect(Collectors.toList());
+								for (ImportDeclaration importDecl : importDecls)
+									for (ImportSpecifier importSpecified : importDecl.getImportSpecifiers()) {
+										if (importSpecified instanceof NamedImportSpecifier) {
+											NamedImportSpecifier namedImportSpecifier = (NamedImportSpecifier) importSpecified;
+											if (namedImportSpecifier.getAlias() != null) {
+												if (namedImportSpecifier.getImportedElement() == selectedElement) {
+													throw new UnsupportedOperationException(
```

Comment:
This is just a temporary way to prevent alias renaming and it only works in certain cases. We need to support alias renaming and get rid of this code .